File: engine/src/strategies/momentum.ts
import type { StrategyModule, SymbolMarketData, Regime, StrategySignal } from './types.js';

function generate(data: SymbolMarketData, regime: Regime): StrategySignal | null {
  if (data.bars.length < 50) return null;

  const recent = data.bars.slice(-10);
  const older = data.bars.slice(-50, -10);
  const recentAvg = recent.reduce((s, b) => s + b.c, 0) / recent.length;
  const olderAvg = older.reduce((s, b) => s + b.c, 0) / older.length;

  const momentum = (recentAvg - olderAvg) / olderAvg;
  // Intraday 1-min bars: 0.01% trend is detectable
  if (Math.abs(momentum) < 0.0001) return null;

  const side = momentum > 0 ? 0 : 1;
  // Scale: 0.02% → 0.4 (Moderate), 0.04% → 0.8 (Strong)
  const conf = Math.min(Math.abs(momentum) * 2000, 0.9);
  const strengthTag = conf >= 0.7 ? 0 : conf >= 0.4 ? 1 : 2;

  return {
    strategy: 3,
    symbol: data.symbol,
    side,
    strength: { TAG: strengthTag, _0: conf },
    target_price: data.price * (1 + momentum * 0.5),
    max_position_pct: regime === 3 ? 0.02 : 0.06,
    timestamp: Date.now(),
  };
}

const momentumStrategy: StrategyModule = {
  id: 3,
  name: 'Momentum',
  minBars: 50,
  generate,
};

export default momentumStrategy;

File: trading-core/lib/signals/momentum.ml
open Trading_core

(** Momentum strategy signal generator.
    EMA 12/26 crossover with 50/200 SMA trend filter, RSI confirmation,
    and relative volume gate. Thresholds configurable via Signal_config. *)

(** Generate a momentum signal.
    @param config      signal configuration (RVOL threshold, RSI levels)
    @param symbol      ticker to trade
    @param ema12       current 12-period EMA
    @param ema26       current 26-period EMA
    @param prev_ema12  previous bar's 12-period EMA
    @param prev_ema26  previous bar's 26-period EMA
    @param sma50       50-period SMA (trend filter)
    @param sma200      200-period SMA (trend filter)
    @param rsi14       14-period RSI value
    @param rel_volume  relative volume (current / average)
    @param price       current price
    @param now         current timestamp *)
let generate ?(config = Signal_config.intraday_defaults.momentum)
    ~symbol ~ema12 ~ema26 ~prev_ema12 ~prev_ema26
    ~sma50 ~sma200 ~rsi14 ~rel_volume ~price ~now () =
  (* Volume gate: configurable RVOL threshold *)
  if rel_volume < config.rvol_threshold then None
  else
    (* Trend filter: 50 SMA above 200 SMA = uptrend, below = downtrend *)
    let uptrend = sma50 > sma200 in
    let downtrend = sma50 < sma200 in
    (* EMA crossover detection *)
    let bullish_cross = prev_ema12 <= prev_ema26 && ema12 > ema26 in
    let bearish_cross = prev_ema12 >= prev_ema26 && ema12 < ema26 in
    (* RSI confirmation *)
    let rsi_oversold = rsi14 < config.rsi_oversold in
    let rsi_overbought = rsi14 > config.rsi_overbought in
    (* Signal logic: crossover in direction of trend with RSI confirmation *)
    let signal_side =
      if bullish_cross && uptrend && not rsi_overbought then Some Types.Buy
      else if bearish_cross && downtrend && not rsi_oversold then Some Types.Sell
      else None
    in
    match signal_side with
    | None -> None
    | Some side ->
      (* Confidence: base 0.6, +0.15 for RSI confirmation, +0.1 for strong RVOL *)
      let base_conf = 0.6 in
      let rsi_bonus =
        match side with
        | Types.Buy -> if rsi_oversold then 0.15 else 0.0
        | Types.Sell -> if rsi_overbought then 0.15 else 0.0
      in
      let vol_bonus = if rel_volume >= config.rvol_strong then 0.1 else 0.0 in
      let conf = Float.min 1.0 (base_conf +. rsi_bonus +. vol_bonus) in
      let strength =
        if conf > 0.8 then Types.Strong conf
        else if conf > 0.5 then Types.Moderate conf
        else Types.Weak conf
      in
      (* Target price: project from EMA difference *)
      let ema_diff = Float.abs (ema12 -. ema26) in
      let target =
        match side with
        | Types.Buy -> price +. ema_diff
        | Types.Sell -> price -. ema_diff
      in
      let sym = match Symbol.create symbol with Ok s -> s | Error _ -> symbol in
      Some ({
        Signal.strategy = Types.Momentum;
        symbol = sym;
        side;
        strength;
        target_price = target;
        max_position_pct = config.max_position_pct;
        timestamp = now;
      } : Signal.t)

(** Check if momentum strategy is enabled for the given phase *)
let is_enabled ~phase =
  Types.strategy_enabled_for_phase phase Types.Momentum
